File: fortran/objshear/test/test-healpix.f90
! vim:set ft=fortran:


program main
    use healpix

    integer*8 nside, n_pix, ipix, i, ira, idec, ringnum
    real*8 area
    real*8 ra(10), dec(13)
    real*8 ra1, dec1, z(4)
    real*8 vector(3)
    real*8 radius
    integer*8 nfound
    integer*8, dimension(:), allocatable :: listpix
    integer lun
    character(255) filename
    
    nside = 4096
    n_pix = npix(nside)
    area = pixarea(nside)
    
    print '("nside: ",i0)',nside
    print '("  npix: ",i0)',n_pix
    print '("  area: ",e)',area*RAD2DEG**2

    ra1=175.0_8
    dec1=27.2_8
    call eq2vec(ra1, dec1, vector)
    print '("convert ",F15.8, F15.8," to vector (",F15.8, F15.8, F15.8,")")',&
        ra1, dec1, vector(1), vector(2), vector(3)

    z=(/-0.75, -0.2, 0.2, 0.75/)
    do i=1,4
        ringnum = ring_num(nside, z(i))
        print '("ring num at z=",F15.8,": ",i0)',z(i),ringnum
    enddo

    ra = (/ 0.,   40.,   80.,  120.,  160.,  200.,  240.,  280.,  320.,  360./)
    dec = (/-90., -85., -65., -45., -25., -5., 0., 5., 25., 45., 65., 85., 90./)

    print '("testing eq2pix")'
    do ira=1,size(ra)
        do idec=1,size(dec)
            call eq2pix(nside, ra(ira), dec(idec), ipix)
            print '(F15.8," ",F15.8," ",i0)',ra(ira),dec(idec),ipix
        end do
    end do

    allocate(listpix(4*nside))
    radius = 40.0_8/3600.0_8*PI/180.0_8

    filename = "test/test-healpix.dat"
    lun=50
    open(unit=lun,file=filename)
    print '("intersect disc")'

    do ira=1,size(ra)
        do idec=1,size(dec)
            call query_disc(nside, ra(ira), dec(idec), radius, listpix, nfound, 1_8)
            print '("  ra: ",F10.6,"  dec: ",F10.6, " pixels:",i0)',ra(ira),dec(idec),nfound

            write (lun,'(F10.6," ",F10.6," ",i0," ",$)'),ra(ira),dec(idec),nfound
            do i=1,nfound
                write (lun,'(i0," ",$)'), listpix(i)
            enddo
            write (lun,'("")')
        end do
    end do

    print '("Closing file: ",a)',filename


end program

File: ccode/objshear/test-healpix.c
#include <stdio.h>
#include <stdlib.h>
#include "healpix.h"
#include "defs.h"

int main(int argc, char** argv) {
    int64 nside=4096;

    struct healpix* hpix = hpix_new(nside);

    printf("nside: %ld\n", nside);
    printf("  npix: %ld\n", hpix->npix);
    printf("  area: %le\n", hpix->area*R2D*R2D);

    double ra1=175.0;
    double dec1=27.2;
    double vector[3];
    hpix_eq2vec(ra1, dec1, vector);
    printf("convert %15.8lf %15.8lf to vector (%15.8lf %15.8lf %15.8lf)\n", 
            ra1, dec1, vector[0], vector[1], vector[2]);

    double z[4]={-0.75, -0.2, 0.2, 0.75};
    for (int i=0; i<4; i++) {
        int64 ringnum = hpix_ring_num(hpix, z[i]);
        printf("ring num at z=%15.8lf: %ld\n", z[i], ringnum);
    }

    double ra[10]={0.,   40.,   80.,  120.,  160.,  200.,  240.,  280.,  320.,  360.};
    double dec[11] = {-85., -65., -45., -25., -5., 0., 5., 25., 45., 65., 85.};
    printf("testing eq2pix\n");
    for (int ira=0; ira<10; ira++) {
        for (int idec=0; idec<11; idec++) {
            int64 ipix = hpix_eq2pix(hpix, ra[ira], dec[idec]);
            printf("%15.8lf %15.8lf %ld\n", ra[ira], dec[idec], ipix);
        }
    }

    hpix_delete(hpix);

    int64 npface = nside*nside;
    printf("ncap from c++: %ld\n", (npface-nside)<<1);
    printf("ncap guess from mine: %ld\n", 2*nside*(nside-1) );
}
